### ISSUE TITLE
does it support `let \p1=(AB) in ...`

### ISSUE THREAD

**Issue by cer-cer** [kevinyuan/vscode-tikz#4]
`
```tikz
\usetikzlibrary{calc}
\begin{document}
\begin{tikzpicture}[scale=2, transform shape]

\draw[step = 1, gray!90, thin] (-3, -3) grid(6, 6);
\draw[->, thick, black] (-3, 0) -- (6, 0) node[above, right]{$X$};
\draw[->, thick, black] (0, -3) -- (0, 6) node[above, left]{$Y$};


\coordinate (A) at (1, 3);
\coordinate (B) at (5, 4);

\draw[red!100, line width=2pt] (A) -- (B);

\coordinate (M) at ($(A)!0.5!(B)$);
\draw[blue, line width=2pt] (M) circle(1 cm);
\node[below] at (M) {$M\circ$};


\coordinate (AB) at ($(B) - (A)$);


\draw[green, line width=1.5pt]
  let \p1=(AB) in 
  ($(M) + 2*(-\y1, \x1)/norm(-\y1, \x1)$) -- ($(M) - 2*(-\y1, \x1)/norm(-\y1, \x1)$);

\end{tikzpicture}
\end{document}
```
`

I tried the code as these, but it output an error:

<img width="772" height="181" alt="Image" src="https://github.com/user-attachments/assets/9f803425-57d1-4ae2-a12c-8d7963a03d94" />

The question:
- does it support the `let`...
- how can i set the `options.showConsole` 

thanks.

**Comment by cer-cer** [comment]
<img width="368" height="232" alt="Image" src="https://github.com/user-attachments/assets/df089aea-9ef3-4c6d-ac67-6889fe09b487" />

**Comment by kevinyuan** [comment]
## Answers

**1. Does `let \p1=... in` work?**

Yes — the `calc` library and `let`/`\p`/`\x`/`\y` syntax are fully supported. The error in your code is that **`norm()` is not a valid TikZ/pgfmath function** — it does not exist in the math engine, which is why it fails. This is not an extension limitation; the same code would fail in a standard pdflatex installation without a custom `norm` definition.

The correct pgfmath function for Euclidean length is `veclen(x, y)`.

**2. How to show console output?**

Add to VS Code settings:

```json
"tikzjax.showConsole": true
```

This prints the TeX engine log to the output panel and makes errors like this one easy to diagnose.

---

## Fix

Replace the `norm()` call with TikZ calc's built-in `!d!angle:` syntax, which computes perpendicular directions without any custom math functions:

```tikz
\usetikzlibrary{calc}
\begin{document}
\begin{tikzpicture}[scale=2, transform shape]

\draw[step = 1, gray!90, thin] (-3, -3) grid(6, 6);
\draw[->, thick, black] (-3, 0) -- (6, 0) node[above, right]{$X$};
\draw[->, thick, black] (0, -3) -- (0, 6) node[above, left]{$Y$};

\coordinate (A) at (1, 3);
\coordinate (B) at (5, 4);

\draw[red!100, line width=2pt] (A) -- (B);

\coordinate (M) at ($(A)!0.5!(B)$);
\draw[blue, line width=2pt] (M) circle(1 cm);
\node[below] at (M) {$M^{\circ}$};

\draw[green, line width=1.5pt]
  ($(M)!2cm!90:(A)$) -- ($(M)!2cm!-90:(A)$);

\end{tikzpicture}
\end{document}
```

`($(M)!2cm!90:(A)$)` means: from M, travel 2 cm in the direction M→A rotated 90° — exactly perpendicular to AB. No `norm()` or manual vector normalization needed. The intermediate `\coordinate (AB)` is no longer needed either.

**Comment by kevinyuan** [comment]
<img width="584" height="559" alt="Image" src="https://github.com/user-attachments/assets/c7b08a95-7bdf-4bb4-9eed-4503d057a3b9" />

**Comment by cer-cer** [comment]
thanks a lot.
is this the error output?

<img width="1647" height="916" alt="Image" src="https://github.com/user-attachments/assets/b0f35bea-5ddb-4324-95b5-91a5f3d4be5f" />
